Typescript types package for client

Would it be possible to have a Typescript types package for client?

Having Typescript types for incoming and outgoing messages could help develop other services based on `zwave-js-server`, for example a plugin for Homebridge.

I'm not sure if they could be generated automatically from the current source code.

I tried, but I had problems with the enums.

Quick example:
```ts
import { IncomingMessage } from "@zwave-js/server/dist/lib/incoming_message.js"
import { WebSocket } from "undici";

export class SocketClient {
    private socket: WebSocket;

    public constructor() {
        this.socket = new WebSocket("ws://192.168.1.5:3000");

        this.send({
            messageId: "initialize",
            command: "initialize",
            schemaVersion: 35
        })
    }

    public send(message: IncomingMessage): void {
        this.socket.send(JSON.stringify(message));
    }
}
```

Typescript error:

src/socketClient.ts:12:13 - error TS2322: Type '"initialize"' is not assignable to type 'ServerCommand | NodeCommand | ControllerCommand | DriverCommand | MulticastGroupCommand | BroadcastNodeCommand | EndpointCommand | UtilsCommand'.

12             command: "initialize",
               ~~~~~~~

  node_modules/@zwave-js/server/dist/lib/incoming_message.d.ts:13:5
    13     command: ServerCommand.startListening;
           ~~~~~~~
    The expected type comes from property 'command' which is declared here on type 'IncomingMessage'


Found 1 error.

Well, for now I am using my own types to incoming/outcoming messages (with several `any`).

But now I'm having problems with the schema version and the `ZwaveState` / `NodeState` types (I imagine `ControllerState` has the same problem, but I'm not using it):

```typescript
import { ZwaveState } from "@zwave-js/server";

protected async handleStart(): Promise<void> {
  await this.sendMessage({
    messageId: "initialize",
    command: "initialize",
    schemaVersion: 35,
  });
  
  const response = await this.sendMessage({
    messageId: "start_listening",
    command: "start_listening",
  });

  this.discoverDevice(response.result.state);
}

private discoverDevice({ nodes }: ZwaveState): void {
  for (const node of nodes) {
    if (node.isControllerNode) {
      continue;
    }
  }
}
```

Any idea how to solve the problem? I can't find a way to select the type corresponding to the correct version of the schema, in this case `35`:
```
src/zwavePlatform.ts:83:22 - error TS2339: Property 'isControllerNode' does not exist on type 'NodeState'.
  Property 'isControllerNode' does not exist on type 'NodeStateSchema0'.

83             if (node.isControllerNode) {
                        ~~~~~~~~~~~~~~~~


Found 1 error.
```

> src/socketClient.ts:12:13 - error TS2322: Type '"initialize"' is not assignable to type 'ServerCommand | NodeCommand | ControllerCommand | DriverCommand | MulticastGroupCommand | BroadcastNodeCommand | EndpointCommand | UtilsCommand'.

I think once https://github.com/zwave-js/zwave-js-server/pull/1259 is merged, you can import `ServerCommand` and use ServerCommand.initialize directly

--- a/src/lib/command.ts
+++ b/src/lib/command.ts
@@ -1,3 +1,12 @@
+export { BroadcastNodeCommand } from "./broadcast_node/command";
+export { ConfigManagerCommand } from "./config_manager/command";
+export { ControllerCommand } from "./controller/command";
+export { DriverCommand } from "./driver/command";
+export { EndpointCommand } from "./endpoint/command";
+export { MulticastGroupCommand } from "./multicast_group/command";
+export { NodeCommand } from "./node/command";
+export { UtilsCommand } from "./utils/command";
+
 export enum ServerCommand {
   startListening = "start_listening",
   updateLogConfig = "update_log_config",

--- a/src/lib/index.ts
+++ b/src/lib/index.ts
@@ -1,6 +1,4 @@
 export * from "./command";
-
-export { ZwavejsServer } from "./server";
 export * from "./state";
-
+export { ZwavejsServer } from "./server";
 export const serverVersion: string = require("../../package.json").version;

--- a/src/lib/index.ts
+++ b/src/lib/index.ts
@@ -1,6 +1,3 @@
-export {} from "./broadcast_node/command";
-export {} from "./config_manager/command";
-
 export {
   BroadcastNodeCommand,
   ConfigManagerCommand,

--- a/src/lib/index.ts
+++ b/src/lib/index.ts
@@ -1,14 +1,4 @@
-export {
-  BroadcastNodeCommand,
-  ConfigManagerCommand,
-  ControllerCommand,
-  DriverCommand,
-  EndpointCommand,
-  MulticastGroupCommand,
-  NodeCommand,
-  ServerCommand,
-  UtilsCommand,
-} from "./command";
+export * from "./command";
 
 export { ZwavejsServer } from "./server";
 export * from "./state";

--- a/src/lib/index.ts
+++ b/src/lib/index.ts
@@ -1,3 +1,18 @@
+export {} from "./broadcast_node/command";
+export {} from "./config_manager/command";
+
+export {
+  BroadcastNodeCommand,
+  ConfigManagerCommand,
+  ControllerCommand,
+  DriverCommand,
+  EndpointCommand,
+  MulticastGroupCommand,
+  NodeCommand,
+  ServerCommand,
+  UtilsCommand,
+} from "./command";
+
 export { ZwavejsServer } from "./server";
 export * from "./state";
 

--- a/src/lib/state.ts
+++ b/src/lib/state.ts
@@ -79,7 +79,7 @@ export interface ControllerStateSchema0 {
   inclusionState: InclusionState;
 }
 
-type ControllerStateSchema16 = Omit<
+export type ControllerStateSchema16 = Omit<
   Modify<
     ControllerStateSchema0,
     {
@@ -90,7 +90,7 @@ type ControllerStateSchema16 = Omit<
   "libraryVersion" | "serialApiVersion"
 >;
 
-type ControllerStateSchema22 = Omit<
+export type ControllerStateSchema22 = Omit<
   Modify<
     ControllerStateSchema16,
     {
@@ -102,28 +102,28 @@ type ControllerStateSchema22 = Omit<
   "isSlave" | "isSecondary" | "isStaticUpdateController"
 >;
 
-interface ControllerStateSchema25 extends ControllerStateSchema22 {
+export interface ControllerStateSchema25 extends ControllerStateSchema22 {
   rfRegion?: RFRegion;
 }
 
-interface ControllerStateSchema31 extends ControllerStateSchema25 {
+export interface ControllerStateSchema31 extends ControllerStateSchema25 {
   status: ControllerStatus;
 }
 
-type ControllerStateSchema32 = Omit<
+export type ControllerStateSchema32 = Omit<
   Modify<ControllerStateSchema31, { isRebuildingRoutes: boolean }>,
   "isHealNetworkActive"
 >;
 
-type ControllerStateSchema34 = ControllerStateSchema32 & {
+export type ControllerStateSchema34 = ControllerStateSchema32 & {
   rebuildRoutesProgress?: ReadonlyMap<number, RebuildRoutesStatus>;
 };
 
-type ControllerStateSchema35 = ControllerStateSchema34 & {
+export type ControllerStateSchema35 = ControllerStateSchema34 & {
   supportsLongRange: MaybeNotKnown<boolean>;
 };
 
-type ControllerStateSchema36 = ControllerStateSchema35 & {
+export type ControllerStateSchema36 = ControllerStateSchema35 & {
   maxLongRangePowerlevel: MaybeNotKnown<number>;
   longRangeChannel: MaybeNotKnown<LongRangeChannel>;
   supportsLongRangeAutoChannelSelection: MaybeNotKnown<boolean>;
@@ -146,42 +146,42 @@ export interface ZwaveState {
   nodes: NodeState[];
 }
 
-interface CommandClassState {
+export interface CommandClassState {
   id: number;
   name: string;
   version: number;
   isSecure: boolean;
 }
 
-interface EndpointStateSchema0 {
+export interface EndpointStateSchema0 {
   nodeId: number;
   index: number;
   installerIcon?: number;
   userIcon?: number;
 }
 
-type EndpointStateSchema3 = Modify<
+export type EndpointStateSchema3 = Modify<
   EndpointStateSchema0,
   { deviceClass: DeviceClassState | null }
 >;
 
-type EndpointStateSchema15 = Modify<
+export type EndpointStateSchema15 = Modify<
   EndpointStateSchema3,
   { commandClasses: CommandClassState[] }
 >;
 
-type EndpointStateSchema26 = Modify<
+export type EndpointStateSchema26 = Modify<
   EndpointStateSchema15,
   { endpointLabel?: string }
 >;
 
-type EndpointState =
+export type EndpointState =
   | EndpointStateSchema0
   | EndpointStateSchema3
   | EndpointStateSchema15
   | EndpointStateSchema26;
 
-interface DeviceClassState0 {
+export interface DeviceClassState0 {
   basic: {
     key: number;
     label: string;
@@ -198,20 +198,20 @@ interface DeviceClassState0 {
   mandatoryControlledCCs: readonly CommandClasses[];
 }
 
-type DeviceClassState36 = Omit<
+export type DeviceClassState36 = Omit<
   DeviceClassState0,
   "mandatorySupportedCCs" | "mandatoryControlledCCs"
 >;
 
-type DeviceClassState = DeviceClassState0 | DeviceClassState36;
+export type DeviceClassState = DeviceClassState0 | DeviceClassState36;
 
-interface ValueState extends TranslatedValueID {
+export interface ValueState extends TranslatedValueID {
   metadata: ValueMetadata;
   ccVersion: number;
   value?: any;
 }
 
-interface ValueMetadataStateSchema0 {
+export interface ValueMetadataStateSchema0 {
   type: ValueType;
   default?: any;
   readable: boolean;
@@ -229,16 +229,16 @@ interface ValueMetadataStateSchema0 {
   unit?: string;
 }
 
-interface ValueMetadataStateSchema28 extends ValueMetadataStateSchema0 {
+export interface ValueMetadataStateSchema28 extends ValueMetadataStateSchema0 {
   stateful?: boolean;
   secret?: boolean;
 }
 
-type ValueMetadataState =
+export type ValueMetadataState =
   | ValueMetadataStateSchema0
   | ValueMetadataStateSchema28;
 
-interface ConfigurationMetadataStateSchema0 {
+export interface ConfigurationMetadataStateSchema0 {
   type: ValueType;
   readable: boolean;
   writeable: boolean;
@@ -262,16 +262,16 @@ interface ConfigurationMetadataStateSchema0 {
   isFromConfig?: boolean;
 }
 
-type ConfigurationMetadataStateSchema29 = Omit<
+export type ConfigurationMetadataStateSchema29 = Omit<
   ConfigurationMetadataStateSchema0,
   "name" | "info"
 >;
 
-type ConfigurationMetadataState =
+export type ConfigurationMetadataState =
   | ConfigurationMetadataStateSchema0
   | ConfigurationMetadataStateSchema29;
 
-interface NodeStateSchema0 extends EndpointStateSchema0 {
+export interface NodeStateSchema0 extends EndpointStateSchema0 {
   status: NodeStatus;
   ready: boolean;
   isListening?: boolean;
@@ -305,12 +305,12 @@ interface NodeStateSchema0 extends EndpointStateSchema0 {
   values: ValueState[];
 }
 
-type NodeStateSchema1 = Modify<
+export type NodeStateSchema1 = Modify<
   NodeStateSchema0,
   { deviceClass: DeviceClassState | null; commandClasses: CommandClassState[] }
 >;
 
-type NodeStateSchema3 = Omit<
+export type NodeStateSchema3 = Omit<
   Modify<
     NodeStateSchema1,
     {
@@ -328,37 +328,40 @@ type NodeStateSchema3 = Omit<
   "maxBaudRate" | "version" | "isBeaming" | "roleType"
 >;
 
-type NodeStateSchema4 = Modify<NodeStateSchema3, { interviewStage?: string }>;
+export type NodeStateSchema4 = Modify<
+  NodeStateSchema3,
+  { interviewStage?: string }
+>;
 
-interface NodeStateSchema5 extends NodeStateSchema4 {
+export interface NodeStateSchema5 extends NodeStateSchema4 {
   deviceDatabaseUrl?: string;
 }
 
-interface NodeStateSchema7 extends NodeStateSchema5 {
+export interface NodeStateSchema7 extends NodeStateSchema5 {
   statistics: NodeStatistics;
 }
 
-interface NodeStateSchema10 extends NodeStateSchema7 {
+export interface NodeStateSchema10 extends NodeStateSchema7 {
   highestSecurityClass: SecurityClass | undefined;
 }
 
-interface NodeStateSchema14 extends NodeStateSchema10 {
+export interface NodeStateSchema14 extends NodeStateSchema10 {
   isControllerNode: boolean;
   keepAwake: boolean;
 }
 
-type NodeStateSchema29 = Omit<NodeStateSchema14, "commandClasses">;
+export type NodeStateSchema29 = Omit<NodeStateSchema14, "commandClasses">;
 
-interface NodeStateSchema30 extends NodeStateSchema29 {
+export interface NodeStateSchema30 extends NodeStateSchema29 {
   lastSeen: MaybeNotKnown<Date>;
 }
 
-interface NodeStateSchema31 extends NodeStateSchema30 {
+export interface NodeStateSchema31 extends NodeStateSchema30 {
   defaultVolume?: number;
   defaultTransitionDuration?: string;
 }
 
-interface NodeStateSchema35 extends NodeStateSchema31 {
+export interface NodeStateSchema35 extends NodeStateSchema31 {
   protocol: Protocols;
 }
 
@@ -376,15 +379,15 @@ export type NodeState =
   | NodeStateSchema31
   | NodeStateSchema35;
 
-interface FoundNodeStateSchema19 {
+export interface FoundNodeStateSchema19 {
   nodeId: number;
   deviceClass: DeviceClassState | null;
   status: NodeStatus;
 }
 
-type FoundNodeStateSchema22 = Omit<FoundNodeStateSchema19, "status">;
+export type FoundNodeStateSchema22 = Omit<FoundNodeStateSchema19, "status">;
 
-type FoundNodeStateSchema23 = Modify<
+export type FoundNodeStateSchema23 = Modify<
   FoundNodeStateSchema22,
   { supportedCCs?: CommandClasses[]; controlledCCs?: CommandClasses[] }
 >;
@@ -394,7 +397,10 @@ export type FoundNodeState =
   | FoundNodeStateSchema22
   | FoundNodeStateSchema23;
 
-function getNodeValues(node: ZWaveNode, schemaVersion: number): ValueState[] {
+export function getNodeValues(
+  node: ZWaveNode,
+  schemaVersion: number,
+): ValueState[] {
   if (!node.ready) {
     // guard: do not request all values (and their metadata) while the node is still being interviewed.
     // once the node hits ready state, all values will be sent in the 'node ready' event.

